### PR TITLE
Reducing compiler warnings

### LINKED ISSUE #2390: Windows: several hundred lines of WARNINGs during build
Windows has a LOT of code-warnings.

We need to fix these, then flip whatever compile option is emitting warnings into emitting errors.

### PR DESCRIPTION
**Pull request by yying**
Closes #2390

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: CMakeLists.txt [hunk 1]
```diff
@@ -260,7 +260,14 @@ if(WINDOWS)
     if(${flag_var} MATCHES "/MD")
       string(REGEX REPLACE "/MD" "/MT" ${flag_var} "${${flag_var}}")
     endif(${flag_var} MATCHES "/MD")
+    
+    # Ignore unused parameter messages for C-based projects (e.g. sqlite3)
+    if (flag_var MATCHES "^CMAKE_C_FLAGS*")
+      set(${flag_var} "${${flag_var}} /wd4101")
```
Comment (yying) [inline CMakeLists.txt]: Currently, we disable warnings for unreferenced variables in sqlite3

File: osquery/config/config.cpp [hunk 2]
```diff
@@ -387,6 +387,8 @@ Status Config::updateSource(const std::string& source,
     json_stream << clone;
     pt::read_json(json_stream, tree);
   } catch (const pt::json_parser::json_parser_error& e) {
```
Comment (theopolis) [inline osquery/config/config.cpp]: Can we instead write `/* e */`?

File: osquery/config/packs.cpp [hunk 3]
```diff
@@ -55,9 +55,10 @@ size_t splayValue(size_t original, size_t splayPercent) {
   }
 
   std::default_random_engine generator;
-  generator.seed(
-      std::chrono::high_resolution_clock::now().time_since_epoch().count());
-  std::uniform_int_distribution<uint32_t> distribution(min_value, max_value);
+  generator.seed(static_cast<unsigned int>(
+      std::chrono::high_resolution_clock::now().time_since_epoch().count()));
+  std::uniform_int_distribution<uint32_t> distribution(
```
Comment (theopolis) [inline osquery/config/packs.cpp]: Can this not do a: `std::uniform_int_distribution<size_t>`?

File: osquery/core/init.cpp [hunk 4]
```diff
@@ -242,7 +242,7 @@ Initializer::Initializer(int& argc, char**& argv, ToolType tool)
       argv_(&argv),
       tool_(tool),
       binary_((tool == ToolType::DAEMON) ? "osqueryd" : "osqueryi") {
-  std::srand(chrono_clock::now().time_since_epoch().count());
+  std::srand(static_cast<unsigned int>(chrono_clock::now().time_since_epoch().count()));
```
Comment (theopolis) [inline osquery/core/init.cpp]: `size_t` doesn't work?

File: osquery/core/init.cpp [hunk 5]
```diff
@@ -514,7 +516,7 @@ void Initializer::initActivePlugin(const std::string& type,
     }
     // The plugin is not local and is not active, wait and retry.
     delay += kExtensionInitializeLatencyUS;
-    sleepFor(kExtensionInitializeLatencyUS / 1000);
+    sleepFor(static_cast<unsigned int>(kExtensionInitializeLatencyUS) / 1000);
```
Comment (theopolis) [inline osquery/core/init.cpp]: Don't we control this method, can't it use a `size_t`?

Comment (yying) [inline osquery/core/init.cpp]: It can

File: osquery/core/init.cpp (hunk 4 above)
Comment (yying) [inline osquery/core/init.cpp]: From my understanding, `std::srand` takes a `unsigned int` so MSVC compiler will complain about truncation.

**Comment by yying** [comment]
Few more errors left:

<pre>
  C:\Users\user\work\osquery\osquery/extensions/interface.h(35): warning C4003: not enough actual parameters for macro 'STR_
EX' [C:\Users\user\work\osquery\build\windows10\osquery\extensions\osquery_extensions.vcxproj]
  C:\Users\user\work\osquery\osquery/extensions/interface.h(36): warning C4003: not enough actual parameters for macro 'STR_
EX' [C:\Users\user\work\osquery\build\windows10\osquery\extensions\osquery_extensions.vcxproj]
  C:\Users\user\work\osquery\osquery/extensions/interface.h(217): warning C4250: 'osquery::extensions::ExtensionManagerHandl
er': inherits 'osquery::extensions::ExtensionHandler::osquery::extensions::ExtensionHandler::ping' via dominance [C:\Users\u
ser\work\osquery\build\windows10\osquery\extensions\osquery_extensions.vcxproj]
  C:\Users\user\work\osquery\osquery/extensions/interface.h(217): warning C4250: 'osquery::extensions::ExtensionManagerHandl
er': inherits 'osquery::extensions::ExtensionHandler::osquery::extensions::ExtensionHandler::call' via dominance [C:\Users\u
ser\work\osquery\build\windows10\osquery\extensions\osquery_extensions.vcxproj]
  C:\Users\user\work\osquery\osquery/extensions/interface.h(35): warning C4003: not enough actual parameters for macro 'STR_
EX' [C:\Users\user\work\osquery\build\windows10\osquery\extensions\osquery_extensions.vcxproj]
  C:\Users\user\work\osquery\osquery/extensions/interface.h(36): warning C4003: not enough actual parameters for macro 'STR_
EX' [C:\Users\user\work\osquery\build\windows10\osquery\extensions\osquery_extensions.vcxproj]
  C:\Users\user\work\osquery\osquery/extensions/interface.h(217): warning C4250: 'osquery::extensions::ExtensionManagerHandl
er': inherits 'osquery::extensions::ExtensionHandler::osquery::extensions::ExtensionHandler::ping' via dominance [C:\Users\u
ser\work\osquery\build\windows10\osquery\extensions\osquery_extensions.vcxproj]
  C:\Users\user\work\osquery\osquery/extensions/interface.h(217): warning C4250: 'osquery::extensions::ExtensionManagerHandl
er': inherits 'osquery::extensions::ExtensionHandler::osquery::extensions::ExtensionHandler::call' via dominance [C:\Users\u
ser\work\osquery\build\windows10\osquery\extensions\osquery_extensions.vcxproj]
</pre>

File: include/osquery/json_parser.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (theopolis) [inline include/osquery/json_parser.h]: Can you move this to `./osquery/core/json.h`? I am resisting adding new files to `./include`.

**Comment by PoppySeedPlehzr** [comment]
ok to test

**Comment by osqueryer** [comment]
:no_good: The commit a9fb24bb1af95dea4b8e749bdfdf770331bbbb0f (Job results: [162](https://jenkins.osquery.io/job/osqueryPullRequestAudit/162/)) failed the code audit.

**Comment by theopolis** [comment]
FYI: The code audit suggestions are valid, no FPs.

**Comment by osqueryer** [comment]
:-1: The commit a9fb24bb1af95dea4b8e749bdfdf770331bbbb0f (Job results: [3179](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3179/)) failed one or more tests (**OS X/Linux**).

**Comment by osqueryer** [comment]
:no_good: The commit f29a485b39be41dab94f791a4f85879c1f4a3949 (Job results: [166](https://jenkins.osquery.io/job/osqueryPullRequestAudit/166/)) failed the code audit.

**Comment by osqueryer** [comment]
:-1: The commit f29a485b39be41dab94f791a4f85879c1f4a3949 (Job results: [3183](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3183/)) failed one or more tests (**OS X/Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 0b4a7f1efbd23176f75ce013c9bd1b8f2b4bf0c4 (Job results: [3185](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3185/)) failed one or more tests (**OS X/Linux**).

**Pull request merged**